File: code_to_improve.py
class Customer:

    def __init__(self, id, n, t):
        self.id = id
        self.n = n
        self.t = t

    def __repr__(self):
        return str(self.id)


class Processor:

    def _get(self, c_list, type):
        r = []
        for i in range(len(c_list)):
            if c_list[i].t == type:
                r.append(c_list[i])
        return r

    def is_paying(self, c):
        if c.t == "standard" or c.t == "premium" or c.t == "enterpise":
            return True
        else:
            return False

    def price(self, c):
        if c.t == "standard":
            return 10.00
        elif c.t == "free":
            return 0
        elif c.t == "premium":
            return 25
        elif c.t == "enterpise":
            return 100

    def process(self, c):
        if self.is_paying(c) == True:
            self.process_premium(c)
        else:
            self.process_freemium(c)

    def process_premium(self, c):
            # Do some service logic here
            print("Paying customer processed")

    def process_freemium(self, c):
        # Do some service logic here
        print("Freemium customer processed")


p = Processor()

c1 = Customer(1, "Bob", "premium")
c2 = Customer(2, "Jane", "standard")
c3 = Customer(3, "Fred", "freemium")
c4 = Customer(4, "Sarah", "premium")
all = [c1, c2, c3, c4]

print("Customer 1 price: " + str(p.price(c1)))
p.process(c1)
print("Premium customers: ", p._get(all, "premium"))

File: code_to_improve.ts
class Customer {
    id: any;
    n: any;
    t: any;

    constructor(id: number, n: any, t: any) {
        this.id = id;
        this.n = n;
        this.t = t;
    }
}

class Processor {
    get(cList: Customer[], type: string): Customer[] {
        const r: Customer[] = [];
        for (let i = 0; i < cList.length; i++) {
            if (cList[i].t === type) {
                r.push(cList[i]);
            }
        }
        return r;
    }

    is_paying(c: Customer): boolean {
        if (c.t == "standard" || c.t == "premium" || c.t == "enterpise") {
            return true
        } else {
            return false
        }
    }

    price(c: Customer): any {
        if (c.t === "standard") {
            return 10.00;
        } else if (c.t === "free") {
            return 0;
        } else if (c.t === "premium") {
            return 25;
        } else if (c.t === "enterpise") {
            return 100;
        }
    }

    process(c: Customer): void {
        if (this.is_paying(c)) {
            this.process_premium(c);
        } else {
            this.process_freemium(c);
        }
    }

    process_premium(c: Customer): void {
        // Do some service logic here
            console.log("Paying customer processed");
    }

    process_freemium(c: Customer): void {
        // Do some service logic here
        console.log("Freemium customer processed");
    }
}

let p = new Processor();

let c1 = new Customer(1, "Bob", "premium");
let c2 = new Customer(2, "Jane", "standard");
let c3 = new Customer(3, "Fred", "freemium");
let c4 = new Customer(4, "Sarah", "premium");
let all: any[] = [c1, c2, c3, c4];

console.log("Customer 1 price: " + p.price(c1));
p.process(c1);
console.log("Premium customers: ", p.get(all, "premium"));

